File: amzn/Linked Lists/golang/ReverseLinkedList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// While you are traversing the list, change the current node's 
// next pointer to point to its previous element. 
// Since a node does not have reference to its previous node, 
//   you must store its previous element beforehand. 
//   You also need another pointer to store the next 
//   node before changing the reference. 
//   Do not forget to return the new head reference at the end!

func reverseList(head *ListNode) *ListNode {
   var  prev *ListNode
    for head != nil {
        temp := head.Next
        head.Next = prev
        prev = head
        head = temp
    }
    return prev
}

File: amzn/Linked Lists/ReverseLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
function reverseList(head) {
    // track previous
    let prev = null;
    while(head!= null){
        // 1 store next so we dont break the link with next step pointing back to previous
        let tmpNext = head.next; 
        // now we can point back 
        head.next = prev;
        // we can bump previous to current node 
        prev = head;
        // bump current to stred next
        head = tmpNext;
    }
    return prev;
}
